src/bin/commands/demux.rs

```diff
@@ -63,6 +59,43 @@ impl ReadSet {
     const COLON: u8 = b':';
     const PLUS: u8 = b'+';
 
+    /// Produces an iterator over references to the template segments stored in this ``ReadSet``.
+    fn template_segments(&self) -> Box<dyn Iterator<Item = &FastqSegment> + '_> {
+        Box::new(self.segments.iter().filter(|s| s.segment_type == SegmentType::Template))
+    }
+    /// Produces an iterator over references to the sample barcode segments stored in this
+    /// ``ReadSet``.
+    fn sample_barcode_segments(&self) -> Box<dyn Iterator<Item = &FastqSegment> + '_> {
+        Box::new(self.segments.iter().filter(|s| s.segment_type == SegmentType::SampleBarcode))
+    }
+    /// Produces an iterator over references to the molecular barcode segments stored in this
+    /// ``ReadSet``.
+    fn molecular_barcode_segments(&self) -> Box<dyn Iterator<Item = &FastqSegment> + '_> {
+        Box::new(self.segments.iter().filter(|s| s.segment_type == SegmentType::MolecularBarcode))
+    }
+
+    /// Generates the sample barcode sequence for this read set and returns it as a Vec of bytes.
+    fn sample_barcode_sequence(&self, sample_barcode_length: usize) -> Vec<u8> {
+        let mut sample_barcode = Vec::with_capacity(sample_barcode_length);
+        for segment in self.sample_barcode_segments() {
+            sample_barcode.extend(segment.seq.clone());
+        }
+        sample_barcode
+    }
+
+    /// Combines ``ReadSet`` structs together into a single ``ReadSet``
+    fn combine_readsets(readsets: Vec<Self>) -> Self {
+        let mut readset_iter = readsets.into_iter();
+        if let Some(mut result) = readset_iter.next() {
+            for next_readset in readset_iter {
+                result.segments.extend(next_readset.segments);
+            }
+            result
+        } else {
+            panic!("Cannot call combine readsets on an empty vec!")
```

Comment:
assert this up front rather than in an else branch at the end.

Reply:
I can do that but I think we'll still need the else block at the end because the compiler isn't smart enough to recognize the assert block on length up front, so it'll still see the else block as reachable.